### ISSUE TITLE
Poor performances for primitive floating-point arithmetic

### ISSUE THREAD

**Issue by silene** [coq/coq#13124]
I have been looking at some performance profiles where functions like `coq_fmul` appear prominently. Yet these functions should not even exist in the first place.

The bytecode interpreter contains assembly instruction such as `call coq_fmul@PLT`, while one would expect a single multiplication opcode there (e.g., `mulsd` on x86). In other words, we are paying a full function call for naught. This might seem like a pessimization from GCC (why not inline the code?), but the compiler is somewhat right there. Indeed the ABI mandates symbol interposition for visible symbols when compiling in PIC mode, and `ocamlc` does pass `-fPIC` to the C compiler.

A simple fix would be to declare `coq_fmul` static, which prevents it from being the target of interposition. Then the compiler performs some inlining, and everything is well. Or not.

Unfortunately, file `kernel/float64.ml` contains the following declaration:
```ocaml
external mul : float -> float -> float = "coq_fmul_byte" "coq_fmul"
[@@unboxed] [@@noalloc]
```
This requires that `coq_fmul` is visible. (Function `Float64.mul` is being used by the standard conversion, the native compiler, and the extraction.)

So, a more complicated fix would be to make two versions of `coq_fmul`, a hidden one that can be used by the bytecode interpreter and a visible one for other uses. In fact, we could just get rid of the hidden one and directly use the floating-point multiplication in `coq_interprete.c`.

That does not solve the performance issue for `native_compute`, though. File `nativevalues.ml` contains the following lines:
```ocaml
let no_check_fmul x y =
  mk_float (Float64.mul (to_float x) (to_float y))
[@@ocaml.inline always]

let fmul accu x y =
  if is_float x && is_float y then no_check_fmul x y
  else accu x y
```

Why do we call `Float64.mul`? Why not use `(*.)` in `no_check_fmul`? Is it because we do not trust the OCaml compiler to produce the correct opcodes?

By the way, note that the `is_float` function used in `fmul` also appears prominently in performance profiles when using `native_compute`. What should have been a single assembly instruction such as `cmpb $-3, -8(%r13)` ends up being a full function call. Short of the OCaml compiler providing an intrinsic for that purpose, I have no idea how to make `native_compute` faster.

**Comment by SkySkimmer** [comment]
>By the way, note that the is_float function used in fmul also appears prominently in performance profiles when using native_compute. What should have been a single assembly instruction such as cmpb $-3, -8(%r13) ends up being a full function call. Short of the OCaml compiler providing an intrinsic for that purpose, I have no idea how to make native_compute faster.

Can we do some sort of check lifting? ie in `fun x y => x * (x * y)` I guess we currently test `is_float x` twice, `is_float y` once and `is_float (x * y)` once but maybe we could compile to something more like
~~~ocaml
fun x y ->
  if is_float x then 
    if is_float y then x * (y * x)
    else (* accu stuff *)
  else (* other accu stuff *)
~~~

**Comment by silene** [comment]
Yes, that is a sensible thing to do. It only works when two primitive operations happen in a row, but it would still give a small speedup. For example, on CoqInterval, this should reduce the number of calls to `is_float` by 33%.

It is not that easy to implement, though. Currently, the compiler is oblivious to the fact that a value might or might not be an accumulator. In other words, the generated code for `x * y + z` is branch-free, something like

    Push x; Push y; CheckMul;
    Push z; CheckAdd;
    Return;

But to make it possible to perform this kind of optimization, it should first be compiled to

    Push x; BranchIfAccu L1; Push y; BranchIfAccu L1; Mul;
    L3: BranchIfAccu L2; Push z; BranchIfAccu L2; Add;
    L4: Return;
    L1: MkAccu Mul; Branch L3;
    L2: MkAccu Add; Branch L4;

which could then be optimized into

    Push x; BranchIfAccu L1; Push y; BranchIfAccu L1; Mul;
    Push z; BranchIfAccu L2; Add;
    L4: Return;
    L1: MkAccu Mul;
    L2: MkAccu Add; Branch L4;

In other words, a lot of knowledge needs to be moved to the compiler about accumulators and primitive operations.

**Comment by silene** [comment]
Pull request #12959 now includes a fix for `vm_compute`.

**Comment by erikmd** [comment]
Thanks @silene!

(Cc @proux01)

regarding your question

> Why do we call `Float64.mul`? Why not use `(*.)` in `no_check_fmul`? Is it because we do not trust the OCaml compiler to produce the correct opcodes?

the motivation for not relying on `( *. )` (and likewise for other float operations) was just to avoid the double-rounding issues triggered when using x87 instructions.

Admittedly, this issue shouldn't occur when using OCaml float ops on a 64-bit architecture, but the choice had been to provide a uniform implementation of the `Float64` module w.r.t. the archs supported by Coq.

However, maybe one might consider doing some conditional compilation here(?)
